node/src/actors/inventory_manager/handlers.rs

```diff
@@ -276,3 +276,188 @@ impl Handler<GetItemSuperblock> for InventoryManager {
+                f().await;
+
+                actix::System::current().stop();
+                Result::<(), ()>::Ok(())
+            };
+
+            // Magic conversion from std::future::Future (futures 0.3) and futures::Future (futures 0.1)
+            let fut = futures_util::compat::Compat::new(Box::pin(fut03));
+
+            actix::Arbiter::spawn(fut);
+        })
+        .unwrap();
+    }
+
+    #[test]
+    fn persist_same_transaction_twice_overwrites() {
+        // Create two blocks with the same transaction, to simulate a reorganization.
+        // GetItemTransaction should return the hash of the last block that was added using
+        // AddItem.
+        test_actix_system(|| async {
+            // Setup testing: use in-memory database instead of rocksdb
+            let mut config = Config::default();
+            config.storage.backend = StorageBackend::HashMap;
+            let config = Arc::new(config);
+            // Start relevant actors
+            config_mngr::start(config);
+            storage_mngr::start();
+            let inventory_manager = InventoryManager::default().start();
+
+            // Create first block with value transfer transactions
+            let block = build_block_with_vt_transactions(1);
+            let block_hash1 = block.hash();
+            let tx_hash1 = block.txns.value_transfer_txns[0].hash();
+            let item = StoreInventoryItem::Block(Box::new(block));
+
+            // Persist first block
+            let f = inventory_manager.send(AddItem { item });
+            let res = Compat01As03::new(f).await.unwrap();
+            res.unwrap();
+
+            // Get first transaction of that block
+            let f = inventory_manager.send(GetItemTransaction { hash: tx_hash1 });
+            let res = Compat01As03::new(f).await.unwrap();
+
+            // The transaction pointer should point to that block
+            let (_tx, tx_pointer1) = res.unwrap();
+            assert_eq!(tx_pointer1.block_hash, block_hash1);
+
+            // Create a different block with the same transactions
+            let block = build_block_with_vt_transactions(2);
+            let block_hash2 = block.hash();
+            let tx_hash2 = block.txns.value_transfer_txns[0].hash();
+            assert_ne!(block_hash1, block_hash2);
+            assert_eq!(tx_hash1, tx_hash2);
+            let item = StoreInventoryItem::Block(Box::new(block));
+
+            // Persist second block
+            let f = inventory_manager.send(AddItem { item });
+            let res = Compat01As03::new(f).await.unwrap();
+            res.unwrap();
+
+            // Get first transaction again
+            let f = inventory_manager.send(GetItemTransaction { hash: tx_hash1 });
+            let res = Compat01As03::new(f).await.unwrap();
+
+            // Now, the transaction pointer should point to the second block
+            let (_tx, tx_pointer2) = res.unwrap();
+            assert_eq!(tx_pointer2.block_hash, block_hash2);
```

Comment:
I am still wondering why this does not work in mainnet, although I guess investigating that becomes another issue

Reply:
I was thinking about adding more tests until we find the issue. The current test checks that the problem is not in `InventoryManager`, so the next step would be to test `ChainManager`, but we need more information to reproduce the issue.

Reply:
I marked issue #1746 as closed because we couldn't reproduce it, I will try to merge this PR anyway because it could be useful in the future, if we make some changes to the `InventoryManager` and want to test it.